[Merged by Bors] - Sync integration

antonlerner opened the pull request:
Integrated fetcher and layer fetcher into current sync flow
test pass

antonlerner committed:
after app test pass

antonlerner committed:
after lint

antonlerner committed:
after fmt

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner committed:
aftet another lint

antonlerner committed:
more fmt

antonlerner committed:
after reverting go get

antonlerner committed:
more go mod nonsense

antonlerner committed:
more and more gomod stuff

antonlerner committed:
go vet fix

antonlerner committed:
fixed hash checking in block fetch

antonlerner committed:
added logs to try debug bugs

antonlerner committed:
added prints to simulator to try and debig

antonlerner committed:
added prints to simulator to try and debig

antonlerner committed:
Merge branch 'sync_integration' of https://github.com/spacemeshos/go-spacemesh into sync_integration

antonlerner committed:
--amend

antonlerner committed:
another tweak

antonlerner committed:
fixed close test and changes atx fetch validation

antonlerner committed:
skipped multi node test

antonlerner committed:
after fmt

antonlerner committed:
fixed failing service test

antonlerner committed:
fixed most sync tests

antonlerner committed:
after lint and fmt

antonlerner committed:
after fmt and vet

antonlerner committed:
fixed last tests

antonlerner committed:
fixed rpc test

antonlerner committed:
removed prints

antonlerner commented:
bors try

antonlerner commented:
bors try

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner commented:
bors try

antonlerner committed:
inifinite retiries to requested hashes

antonlerner committed:
Merge branch 'sync_integration' of https://github.com/spacemeshos/go-spacemesh into sync_integration

antonlerner commented:
bors try

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner commented:
bors try

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner committed:
working version

y0sher committed:
set layers config timeout.

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner committed:
merge fixes

antonlerner committed:
Merge branch 'sync_integration' of https://github.com/spacemeshos/go-spacemesh into sync_integration

antonlerner committed:
after fmr

antonlerner commented:
bors merge

antonlerner committed:
removed commented code

antonlerner committed:
Merge branch 'develop' into sync_integration

y0sher commented on activation/atxdb.go:
```diff
@@ -524,7 +524,8 @@ func (db *DB) storeAtxUnlocked(atx *types.ActivationTx) error {
 
-	atxBodyBytes, err := types.InterfaceToBytes(getAtxBody(atx))
+	// todo: this changed so that a full atx will be written - inherently there will be double data with atx header
+	atxBodyBytes, err := types.InterfaceToBytes(atx)
```
so maybe let's rename as this is the full atx bytes and not just body..

y0sher replied:
and we can remove the comment below

noamnelke replied:
why was it necessary to write the full ATX and not store the header separately?

antonlerner replied:
Since `ActivationTxHeader` is public it is written to DB anyways as zeros +
New sync paradigm is to store data with it's respected hash as key - so ATX needs to be full for validation

y0sher commented on activation/atxdb.go:
```diff
@@ -775,14 +778,14 @@ func (db *DB) HandleAtxData(data []byte, syncer service.Fetcher) error {
 	if err := syncer.GetPoetProof(atx.GetPoetProofRef()); err != nil {
-		return fmt.Errorf("received atx (%v) with syntactically invalid or missing PoET proof (%x): %v",
-			atx.ShortString(), atx.GetShortPoetProofRef(), err)
+		return fmt.Errorf("received ATX (%v) with syntactically invalid or missing PoET proof (%x): %v",
```
I've seen a PR from @lrettig where he changes all to lowercase `atx` , we probably need to decide on a convention..

lrettig replied:
yes, I prefer all lowercase, it makes searching/grepping through logs easier - definitely lmk if you guys disagree

y0sher commented on activation/atxdb.go:
```diff
@@ -822,3 +825,16 @@ func (db *DB) FetchAtxReferences(atx *types.ActivationTx, f service.Fetcher) err
 
 	return nil
 }
+
```
remove?

y0sher commented on cmd/node/multi_node_test.go:
```diff
@@ -5,5 +5,6 @@ import (
 )
 
 func TestMultiNode(t *testing.T) {
+	t.Skip()
```
bring back?

y0sher commented on fetch/fetch.go:
```diff
@@ -128,26 +141,31 @@ func NewMessageNetwork(requestTimeOut int, net service.Service, protocol string,
+func GetRandomPeer(peers []p2ppeers.Peer) p2ppeers.Peer {
 	if len(peers) == 0 {
-		f.Log.Panic("cannot send fetch - no peers found")
+		log.Panic("cannot send fetch - no peers found")
```
should we really panic here?

antonlerner replied:
what should we do if we try to get a peer and dont have one?

noamnelke commented on activation/atxdb.go:
```diff
@@ -720,11 +721,13 @@ func (db *DB) GetFullAtx(id types.ATXID) (*types.ActivationTx, error) {
 	if err != nil {
 		return nil, err
 	}
-	header, err := db.GetAtxHeader(id)
+	/*header, err := db.GetAtxHeader(id)
 	if err != nil {
 		return nil, err
 	}
 	atx.ActivationTxHeader = header
+
+	*/
```
This change is problematic - the resulting ATX will have no ID set. The `GetAtxHeader` method sets the ID field in the header, which is a private field (lowercase) so isn't stored to disk.

antonlerner replied:
setting stuff in getter usually leads to these kind of problems... we should think about a way to solve this (I can't think of how right now)

noamnelke replied:
the getter is responsible for deserializing the header and returning an initialized Go structure. The ID field isn't stored on disk since it's essentially just a cache. Setting it in the getter avoids the need to recalculate the ID (which is a hash of the ATX) when we already know it (that's the key of the database entry).

I don't think that the way it was implemented is problematic and I don't see the motivation for this change here. Maybe if I understand the motivation I can suggest a solution.

y0sher commented on blocks/blocks.go:
```diff
@@ -200,8 +200,11 @@ func (bh *BlockHandler) fetchAllReferencedAtxs(blk *types.Block, syncer service.
-	err := syncer.GetAtxs(atxs)
-	return err
+	if len(atxs) > 0 {
+		err := syncer.GetAtxs(atxs)
```
```suggestion
		return syncer.GetAtxs(atxs)
```

y0sher replied:
and remove the line below

y0sher commented on blocks/blocks.go:
```diff
@@ -200,8 +200,11 @@ func (bh *BlockHandler) fetchAllReferencedAtxs(blk *types.Block, syncer service.
-	return err
+	if len(atxs) > 0 {
+		err := syncer.GetAtxs(atxs)
+		return err
```
```suggestion
		
```

y0sher commented on cmd/node/node.go:
```diff
@@ -575,7 +580,18 @@ func (app *SpacemeshApp) initServices(nodeID types.NodeID,
+	}
+	blockListener := blocks.NewBlockHandler(bCfg, msh, eValidator, lg)
+
+	fetcher := fetch.NewFetch(app.Config.FETCH, swarm, app.addLogger(Fetcher, lg))
```
nameing is kinda weird here and also in general,  - you are using `fetch` and `fetcher` alternately in a way that doesn't  really make sense.

y0sher commented on cmd/node/node.go:
```diff
@@ -644,6 +654,8 @@ func (app *SpacemeshApp) initServices(nodeID types.NodeID,
 	app.oracle = blockOracle
 	app.txProcessor = processor
 	app.atxDb = atxdb
+	app.fetch = fetcher
```
if `fetch` isn't `Start'ed or `Close`ed maybe we don't have to keep it in the `app` struct?

y0sher commented on collector/memory_collector.go:
```diff
@@ -171,3 +171,11 @@ func (c *MemoryCollector) GetReceivedBlocks(layer types.LayerID) int {
+
+// AtxIDExists checks if an ATX with a certain ID exists.
+func (c *MemoryCollector) AtxIDExists(atxID string) bool {
+	c.lck.Lock()
```
can't it be RLock?

y0sher commented on fetch/fetch.go:
```diff
@@ -227,32 +260,34 @@ func (f *Fetch) loop() {
 			f.activeReqM.Unlock()
+			f.log.Info("request added to queue %v", req.hash.ShortString())
 			if req.priority > Low {
 				f.sendBatch([]requestMessage{{req.hint, req.hash}})
```
I understand we send high priority requests right away.. we're purposely not batching them?

antonlerner replied:
this was the initial thought yes
I can change that to batch some requests

y0sher commented on fetch/fetch.go:
```diff
@@ -367,6 +399,7 @@ func (f *Fetch) requestHashBatchFromPeers() {
 	var requestList []requestMessage
 	f.activeReqM.RLock()
 	for hash, req := range f.activeRequests {
+		f.log.Info("batching %v", hash.ShortString())
```
maybe write a single log line specifying all batched hashes instead of this? maybe move this debug..

y0sher commented on fetch/fetch.go:
```diff
@@ -227,32 +260,34 @@ func (f *Fetch) loop() {
 				break
 			}
 			if rLen > batchMaxSize {
-				f.requestHashBatchFromPeers() // Process the batch.
```
should we limit concurrency here? at least make sure `requestHashBatchFromPeers` will never hang after `stop` channel got a message.

y0sher commented on mesh/mesh.go:
```diff
@@ -487,6 +499,32 @@ func (msh *Mesh) getRunningLayerHash(layerID types.LayerID) (types.Hash32, error
+func (msh *Mesh) GetLayerHashBlocks(h types.Hash32) []types.BlockID {
+	layerIDBytes, err := msh.general.Get(h.Bytes())
+	if err != nil {
+		msh.Warning("requested unknown layer hash %v", h.Hex())
```
why not return error? otherwise you won't be able to know empty layer and not existing layer

antonlerner replied:
opened #2347

y0sher commented on sync/fetch_queue_test.go:
```diff
@@ -41,11 +41,12 @@ func SyncFactory(name string, serv service.Service) *Syncer {
 }
 
 func TestBlockListener_TestTxQueue(t *testing.T) {
+	t.Skip()
```
fix tests or remove before merging ?

antonlerner replied:
I'd rather remove it as a whole in another PR, just want this in already

y0sher commented on sync/syncer.go:
```diff
@@ -452,11 +475,16 @@ func (s *Syncer) handleNotSynced(currentSyncLayer types.LayerID) {
 		}
 
-		lyr, err := s.getLayerFromNeighbors(currentSyncLayer)
+		lyr, err := s.getLayerFromNeighbors2(currentSyncLayer)
```
lets remove old `getLayerFromNeighbors` and refactor this to be it

y0sher commented on sync/syncer.go:
```diff
@@ -329,33 +345,40 @@ func (s *Syncer) run() {
+			s.Log.Info("start syncing epoch atxs")
+			err := s.fetcher.GetEpochATXs(curr.GetEpoch()) //syncEpochActivations(curr.GetEpoch())
+			if err != nil {
+				if curr.GetEpoch().IsGenesis() {
```
should we use `GetEffectiveGenesis` here instead?

antonlerner replied:
no, in genesis epochs this shouldn't be an error - in non genesis atxs are generated and therefore this should be an error

y0sher commented on sync/syncer.go:
```diff
@@ -765,8 +805,9 @@ func (s *Syncer) fetchAllReferencedAtxs(blk *types.Block) error {
-	_, err := s.atxQueue.HandleAtxs(atxs)
-	return err
+	return s.fetcher.GetAtxs(atxs)
+	//_, err := s.atxQueue.HandleAtxs(atxs)
```
remove old comments

y0sher commented on sync/syncer.go:
```diff
@@ -1164,7 +1205,8 @@ func (s *Syncer) FetchPoetProof(poetProofRef []byte) error {
 func (s *Syncer) GetPoetProof(hash types.Hash32) error {
-	poetProofRef := hash.Bytes()
+	return s.fetcher.GetPoetProof(hash)
+	/*poetProofRef := hash.Bytes()
```
remove comments

antonlerner committed:
after CR

antonlerner committed:
Merge https://github.com/spacemeshos/go-spacemesh into sync_integration

antonlerner committed:
Merge branch 'sync_integration' of https://github.com/spacemeshos/go-spacemesh into sync_integration

antonlerner committed:
fixed the merge problems

antonlerner committed:
fixed lint

antonlerner committed:
fixed test

antonlerner committed:
skipped multi node

antonlerner commented:
bors try

antonlerner commented:
bors try

antonlerner committed:
fixed multi node sim and the sync test

antonlerner committed:
Merge branch 'develop' into sync_integration

y0sher committed:
Sync tortoise integration  (#2373)

## Motivation
this merges the new sync code with the new verifying tortoise code. 

TODO: generate new sync database,

Co-authored-by: anton <anton.lerner@gmail.com>

y0sher reviewed (approved):
Looks good, happy that we can finally merge this

noamnelke commented on activation/atxdb.go:
```diff
@@ -524,11 +524,12 @@ func (db *DB) storeAtxUnlocked(atx *types.ActivationTx) error {
 	}
 
-	atxBodyBytes, err := types.InterfaceToBytes(getAtxBody(atx))
+	// todo: this changed so that a full atx will be written - inherently there will be double data with atx header
```
@antonlerner what's the advantage of storing the header fields twice? one instead of two db hits when syncing?

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner committed:
fixed some tests

antonlerner committed:
after merge fixes

antonlerner committed:
after fmt

antonlerner committed:
removed not necessary fix

antonlerner committed:
removed -v since I suspect this what cuases tests to run so slow

antonlerner committed:
fixed p2p merg error

antonlerner committed:
fixed tmp dir in node_test

antonlerner committed:
after fmt

antonlerner committed:
added prints to p2p test for debugging, better safer way to initialize node

antonlerner commented:
bors merge

antonlerner commented:
bors ping

antonlerner commented:
bors p=10000

antonlerner commented:
bors r-

antonlerner commented:
bors r+

antonlerner committed:
fixed flaky sync test

antonlerner commented:
bors merge

antonlerner commented:
bors merge

antonlerner committed:
fixed stuck sync test

antonlerner commented:
bors merge

antonlerner committed:
fixed activation stalling close

antonlerner commented:
bors merge

antonlerner commented:
bors merge

antonlerner committed:
reduced p2p test timeout

antonlerner commented:
bors merge

antonlerner commented:
bors mrege

antonlerner commented:
bors r+

antonlerner committed:
removed p2p small integration test to see if this affects overall tests

antonlerner commented:
bors merge

antonlerner committed:
fixed pipline config

antonlerner commented:
bors r+

antonlerner committed:
added -v flag to see where tests hang

antonlerner committed:
attempt to fix sync test

antonlerner commented:
bors r+

antonlerner committed:
fixed clock leaks in sync tests

antonlerner committed:
after fmt

antonlerner commented:
bors r+

y0sher committed:
fix wrong variable naming that rarely created infinite loop

antonlerner commented:
bors r+

y0sher committed:
Merge branch 'p2p_hanging_testbugfix' into sync_integration

antonlerner commented:
bors r+

antonlerner committed:
Merge branch 'develop' into sync_integration

antonlerner commented:
bors r+

antonlerner commented:
bors merge

antonlerner committed:
removed a lot of logs

antonlerner committed:
fixed nil ctx in sync mock constructor

antonlerner commented:
bors merge

antonlerner committed:
fixed concurrency in sync test

antonlerner commented:
bors merge

antonlerner committed:
temp disabled new debug messages

antonlerner commented:
bors r+
